### ISSUE TITLE
NotificationThreads's groupDevices method throwing IllegalArgumentException

### ISSUE THREAD

**Issue by mwkirk** [mwkirk/javapns#71]
_Original author: ypf327...@gmail.com (September 30, 2011 10:06:26)_

hi, there is something wrong at &quot;javapns.notification.transmission.NotificationThreads&quot;
...
private List&lt;List&lt;Device&gt;&gt; groupDevices(List&lt;Device&gt; devices, int threads);

/////////////////////////////////////////////
public class Test {

```
public static void main(String[] args) throws Exception {

    for (int n = 10; n &lt; 10000; n++) {
        List&lt;Device&gt; devices = new ArrayList&lt;Device&gt;();
        for (int i = 0; i &lt; n; i++) {
            devices.add(new BasicDevice(&quot;123456789012345678901234567890123456789012345678901234567&quot; + (1000000 + i)));
        }
        try {
            groupDevices(devices, 10);
        } catch (IllegalArgumentException e) {
            e.printStackTrace();
            System.out.println(&quot;............n=&quot; + n);
        }
    }
}

private static List&lt;List&lt;Device&gt;&gt; groupDevices(List&lt;Device&gt; devices, int threads) {
    List&lt;List&lt;Device&gt;&gt; groups = new Vector&lt;List&lt;Device&gt;&gt;();
    int total = devices.size();
    int devicesPerThread = (total / threads);
    if (total % threads &gt; 0) devicesPerThread++;
    for (int i = 0; i &lt; threads; i++) {
        int firstDevice = i * devicesPerThread;
        int lastDevice = firstDevice + devicesPerThread - 1;
        if (lastDevice &gt;= total) lastDevice = total - 1;
        List&lt;Device&gt; threadDevices = devices.subList(firstDevice, lastDevice + 1);
        groups.add(threadDevices);
    }
    return groups;
}
```

}

//////////////////////////////////////////////////////////

See the result above you will find the problem.

_Original issue: http://code.google.com/p/javapns/issues/detail?id=71_

**Comment by mwkirk** [comment]
_From sype...@gmail.com on September 30, 2011 14:59:14_
Thank you for reporting the issue.  It was reproduced with the test case you provided and a fix will be issued shortly.

**Comment by mwkirk** [comment]
_From sype...@gmail.com on September 30, 2011 22:21:55_
Fixed in 2.0 RC Revision 001, which you can download as of now from the source:
http://code.google.com/p/javapns/source/browse/#svn%2Fbranches%2Fjavapns2

**Comment by mwkirk** [comment]
_From ypf327...@gmail.com on October 08, 2011 01:44:41_
thanks for your efforts
